kms/pkcs11/pkcs11.go

```diff
@@ -70,6 +71,14 @@ func New(ctx context.Context, opts apiv1.Options) (*PKCS11, error) {
 			}
 			config.SlotNumber = &n
 		}
+		// Get module or default to use p11-kit-proxy.so.
+		//
+		// pkcs11.New(module string) will use dlopen that will look for the
+		// given library in the appropriate paths, so there's no need to provide
+		// the full path.
+		if config.Path = u.Get("module-path"); config.Path == "" {
+			config.Path = defaultModule
+		}
```

Comment:
This is in the wrong place, isn't it?  It'll only happen if a URI is given, but there's an alternative code flow where it's all in the config? That should benefit from the default, too.

There's already a check for `config.Path == ""` a few lines further down, with an error return which is no longer telling the truth because it says that 'module-path is required'.

Also... what's wrong with `pkcs11:manufacturer=piv_II;id=%01` which I usually use as my examples in documentation? Why are only token/serial/slot-id accepted, and not model/manufacturer? 

(And why is this something you even need to do for yourself; surely "match token against the attributes in a URI" is a generic thing that ought to be somewhere much further down the stack than your code; in ThalesIgnite/crypto11 or miekg/pkcs11. Ideally you just want to call a function like the [matchSlots()](https://github.com/dwmw2/rolesanywhere-credential-helper/blob/ced1c51d62aeae24198ce489e02180c5ae1a025b/aws_signing_helper/pkcs11_signer.go#L66) I wrote here, and in your case as discussed, you want to just bail if `len(slots) != 1`).

Reply:
> I'd still like to see the documentation at least mention that in a properly configured system, module-path should be unnecessary.

In particular, these two:
 * https://smallstep.com/docs/step-ca/configuration/#pkcs-11
 * https://github.com/smallstep/step-kms-plugin/blob/main/README.md
 
 .. especially the former. I don't want to live in a world where we have to tell users to explicitly pass a `module-path`.
  
 All those examples for how to use different HSMs differ mostly in the `module-path`, which shouldn't be needed anyway.
 
The packaging of the PKCS#11 provider module itself should come with a p11-kit .module file, the token should automatically appear in the output of things like `p11-kit list-tokens` or `p11tool --list-tokens` _(and thus also in GUIs  when you're selecting a client cert for 802.1x or VPN auth, etc.)_
 
 The common case is that user should just be able to select a token from the list, e.g. `pkcs11:manufacturer=piv_II`.

Reply:
> This is in the wrong place, isn't it? It'll only happen if a URI is given, but there's an alternative code flow where it's all in the config? That should benefit from the default, too.

PKCS#11 cannot be configured without a URI, currently you will get the following error:
```
return nil, errors.New("kms uri 'module-path' are required")
```

We can make it more explicit and actually fail if URI == "".

> Also... what's wrong with pkcs11:manufacturer=piv_II;id=%01 which I usually use as my examples in documentation? Why are only token/serial/slot-id accepted, and not model/manufacturer?

Nothing, but https://github.com/ThalesIgnite/crypto11, which is the one we use on top of github.com/miekg/pkcs11 doesn't support it, see [this code](https://github.com/ThalesIgnite/crypto11/blob/a81014c7c41025fb5533c0c6b1b14bec016be695/crypto11.go#L285-L298).

Regarding the documentation, some of it is managed by a different team, and we haven't integrated this yet, so it will have to wait a bit. But I can change `step-kms-plugin` README and help once this is integrated.

Reply:
I've fixed some of this in a23378d